쥬스 메이커 [STEP 3] Lingo, Taeangel

seu0313 opened the pull request:
@jungseungyeo 
안녕하세요 린생 😃
9_group의 Lingo, Taeangel 입니다. 잘 부탁드립니다!
쥬스 메이커 STEP 3 PR 보냅니다.

## 고민한 점
1. JuiceViewController에 있는 juiceMaker를 StockViewController에 어떻게 전달할지에 대해 고민했습니다.
2. IBOutlet을 각각 하나씩 선언해줘 사용하기 불편한 것을 어떻게 개선할 지 고민했습니다.
3. LandScape 모드 (가로모드)에서 Modal Page Sheet와 Full Screen 차이에 대해 고민했습니다.
4. StockViewController의 navigationItem과 NavigationController의 navigationItem의 차이에 대해 고민했습니다.

## 해결한 점
1. IBOutlet을 각각 선언해줬는데 IBOutlet Collection을 사용해 IBOutlet 배열과 tag를 통해 해결했습니다.
2. UIViewController에 navigationItem 프로퍼티가 있고 StockViewController와 NavigationController는 UIViewController를 상속받고 있기 때문에 각각 navigationItem을 사용할 수 있는 것이고 각각 다른 것을 보게된다는 점을 알게 되었습니다.
3. LandScape 모드 (가로모드)에서 Modal Page Sheet와 Full Screen 차이는 
Page Sheet에서 전체 화면으로 덮어지지만 사실 첫번째 ViewController가 사라지지 않고 두번째 ViewController가 덮여지는 방식이고
Full Screen일때는 첫번째 ViewController가 사라지면서 두번째 ViewController가 나오는 방식이라는 것을 View Life Cycle 출력을 통해 알았습니다.

## 조언 받고 싶은 점
1. 아래의 코드와 같이 길어진 guard 문의 조건에서 저희는 대입 연산자를 기준으로 띄우는 방식과 파라미터를 띄우는 방식으로 해결했는데 더 좋은 방법이 있을 지 궁금합니다.
```swift
func presentStockViewController() {
  guard let navigationController =
          self.storyboard?.instantiateViewController(
            withIdentifier: StoryboardID.stockNavigationController) as? UINavigationController,
        let stockViewController =
          navigationController.visibleViewController as? StockViewController
  else {
    return
  }
  stockViewController.juiceMaker = juiceMaker
  stockViewController.delegate = self
  self.present(navigationController, animated: true)
}
```

2. 기존의 String으로 변경해주는 함수를 updateLabel() 로 다시 합쳤는데 이렇게 작성하는 것이 좋은 코드일지 기존대로 나누는 것이 좋을 지 궁금합니다.
```swift
func updateLabel() {
  for (label, fruit) in zip(fruitCountLabels, Fruit.allCases) {
    if let fruitAmount = self.juiceMaker?.fruitStore.stock[fruit] {
      label.text = String(fruitAmount)
    } else {
      label.text = String(Int.zero)
    }
  }
}
```

Taeangel committed:
feat: StockViewController의 재고 추가 및 닫기버튼 구현

화면 전환방식을 modal방식으로 하여 닫기버튼을 구현

Commit:
refactor: 화면 전환 방식을 스토리보드애서 코드 present 방식으로 변경

화면 전환 방식을 스토리보드의 Segue애서 didTapChangeStockButton을 사용하여
코드로 present 하는 방식으로 변경 및 StockViewController의 juiceMaker 프로퍼티 주입
각 스토리보드의 ViewController 들의 ID를 enum 상수로 관리

Taeangel committed:
feat: Delegation pattern 적용

StockViewController의 delegate를 JuiceViewController로 작업을 위임

Commit:
feat: StockViewController에 updateLabel, updateStepper 함수 구현

StockViewController에 update와 convert 함수를 생성하여 JuiceMaker가 가지고 있는
FruitStore의 stock 데이터와 Label과 Stepper를 연결

Commit:
refactor: IBOutlet Collection으로 IBOutlet를 묶음, 중복된 코드 리팩토링

IBOutlet Collection으로 IBOutlet들을 배열로 묶고 zip 함수를 사용해서 각 UI 요소를 변경
sender의 tag를 사용하기 위해 Fruit의 rawValue를 Int로 설정
convert 기능을 각 update 함수에서 처리하도록 변경

Taeangel committed:
feat: 오토레이아웃 설정

메인스토리보드의 StockViewController에 오토레이아웃을 적용

Commit:
style: 메인 스토리보드 기본 화면 기종 변경

iPhone 4s 에서 iPhone 13 Pro로 변경

Commit:
style: 누락된 IBOutlet Collection private 키워드 수정

IBOutlet에서 IBOutlet Collection으로 변경하는 과정에서 누락된 private 키워드 수정

jungseungyeo commented on JuiceMaker/JuiceMaker/Controller/StockDelegate.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+protocol StockDelegate {
```
여기서 AnyObject를 채택해주면 좋을거같아요! 이걸 채택하는 이유를 고민해주세요 :)

Taeangel replied:
기존의 프로토콜은 클래스와 구조체 열거형 모두 채택할 수 있지만 AnyObject를 채택할 경우는 클래스만 채택하게 하여 
```swift
weak var delegate: StockDelegate?
```
이처럼 사용하여 약한참조를 걸기 위해 AnyObject를 채택합니다.

jungseungyeo commented on JuiceMaker/JuiceMaker/Controller/StockViewController.swift:
```diff
@@ -9,23 +9,56 @@ import UIKit
   
-  @IBAction private func didStepperValueChanged(_ sender: UIStepper) {}
+  func updateLabel() {
+    for (label, fruit) in zip(fruitCountLabels, Fruit.allCases) {
```
zip을 사용한거 너무 좋네요 전 step부터 오퍼레이터를 잘 사용하는 모습 좋습니다 :)

jungseungyeo commented on JuiceMaker/JuiceMaker/Util/Constant/StoryboardID.swift:
```diff
@@ -0,0 +1,14 @@
+import Foundation
+
+enum StoryboardID {
+  static let main = "Main"
```
이렇게 상수를 관리하면 좋지만, 파일과 네이밍을 관리하는데 휴먼 리소스가 좀 들어가지 않나요? 어떻게 생각하시나요?

Taeangel replied:
ViewController의 id를 관리하던 폴더를 삭제하고 `StockNavigationController`의 `static let identifier = String(describing: StockNavigationController.self)`를 만들어서
```swift
withIdentifier: StockNavigationController.identifier
```
이와같이 접근할 수 있게 변경하였습니다.

seu0313 replied:
확실히 추가될 때마다 문자열로 하나씩 적어주는 작업에서 휴먼 리소스가 들어갈 것 같습니다.
`JuiceViewController`나 `StockViewController` 인 경우 따로 클래스 소스코드가 있기 때문에 아래와 같이 구현이 가능하다고 생각합니다.

```swift
final class StockViewController: UIViewController {
  static let identifier = String(describing: StockViewController.self)
}
```
하지만, StockNavigationController와 같이 스토리보드 상에서만 존재하는 경우
리팩토링된 것처럼 따로 StockNavigationController 소스 코드를 작성해주는 것이 더 좋을 지 고민입니다.

jungseungyeo commented on JuiceMaker/JuiceMaker/Controller/JuiceViewController.swift:
```diff
@@ -87,13 +74,26 @@ private extension JuiceViewController {
+            self.storyboard?.instantiateViewController(
+              withIdentifier: StoryboardID.stockNavigationController) as? UINavigationController,
+          let stockViewController =
+            navigationController.visibleViewController as? StockViewController
```
질문했던 내용이군요 ㅎㅎ 좋습니다

jungseungyeo commented on JuiceMaker/JuiceMaker/Controller/StockViewController.swift:
```diff
@@ -9,23 +9,56 @@ import UIKit
-  @IBOutlet private weak var kiwiStepper: UIStepper!
-  @IBOutlet private weak var mangoStepper: UIStepper!
+  var juiceMaker: JuiceMaker?
+  var delegate: StockDelegate?
```
delegate패턴을 쓰셨군요 ㅎㅎ 너무 좋습니다 하지만 조금 수정하면 좋을거같아요 

`weak var delegate: StockDelegate?`

위와 같은 코드로 변경을 해주세요 이렇게 하는 이유는 무엇일까요?

seu0313 replied:
프로토콜은 구조체, 열거형, 클래스 모두에서 채택할 수 있지만 `AnyObject`를 채택할 경우 클래스만 가능하도록 제한할 수 있기 때문에 참조 타입이 될 것이라는 것을 명시할 수 있게 됩니다. 

`JuiceViewController` 와 `StockViewController` 가 서로를 강한 참조를 하고 있을 경우 RC가 각각 1 씩 더 증가하기 때문에 화면에서 dismiss 되더라도 메모리에서는 서로를 참조하고 있는 상태로 남게 되어 **메모리 누수**가 발생합니다.

 이때 `weak` 키워드를 통해 약한 참조를 하게되면 RC가 증가하지 않게되어 (소유하지 않게되어) **메모리 누수**를 막을 수 있기 때문에 아래와 같이 사용하게 됩니다.
```swift
weak var delegate: StockDelegate?
```

jungseungyeo commented on JuiceMaker/JuiceMaker/Controller/JuiceViewController.swift:
```diff
@@ -87,13 +74,26 @@ private extension JuiceViewController {
       return
     }
-    self.present(stockViewController, animated: true)
+    stockViewController.juiceMaker = juiceMaker
```
이렇게 직접 넣어주면 캡슐화의 위배가 되는거 같아요 ㅎㅎ 다른 방법이 있을거같아요

seu0313 replied:
프로퍼티 주입을 하게 되면 캡슐화의 위배가 되기 때문에 `setJuiceMaker` 함수를 구현하여 `StockViewController` 내에 있는 juiceMaker를 외부에서 접근할 수 없도록 하였습니다.

jungseungyeo reviewed (changes requested):
링고, 태엔젤 고생하셨어요 ㅎㅎ 
점점 코드가 좋아지네요 ㅎㅎ 고생하셨습니다. 
질문의 답변을 드리겠습니다.

1. 아래의 코드와 같이 길어진 guard 문의 조건에서 저희는 대입 연산자를 기준으로 띄우는 방식과 파라미터를 띄우는 방식으로 해결했는데 더 좋은 방법이 있을 지 궁금합니다.

> 물론있죠 ㅎㅎ 지금 생성하는 곳이 JuiceViewController에서 StockViewController의 인스턴스를 생성하고 있습니다. 이런 경우 인스터스를 누가 제공하면 좋을까요? 자기 자신의 인스턴스는 자기가 만들어서 제공해주는 것이 좋아요. StockViewController에서 제공해준다면 이 함수는 충분히 코드의 량이 줄어들거같아요.

2. 기존의 String으로 변경해주는 함수를 updateLabel() 로 다시 합쳤는데 이렇게 작성하는 것이 좋은 코드일지 기존대로 나누는 것이 좋을 지 궁금합니다.

> 저는 개인적으로는 전의 코드가 좋지만 지금 코드가 나쁘다는 의미는 아닙니다. 이건 개인의 선택 부분인거같아요. 저는 변경 된 코드도 너무 좋습니다. 특히 zip를 활용하여 update를 하는 것이 인상적이였습니다.

링고, 태엔젤 코드를 보면서 리뷰 할 곳이 없을 정도로 잘 하셨어요. 그래도 몇 군데 리뷰를 남겼습니다. 리뷰를 보시고 수정해주시면 좋을거같아요. 

궁금한 점이 있으면 언제든지 물어봐주세요 :)

Taeangel committed:
refactor: StockDelegate에 AnyObject 채택, delegate의 강한참조를 약한참조로 변경

StockDelegate에 AnyObject 클래스만 채택할 수 있게 채택하고 weak을 사용하여 delegate의
강한참조를 약한참조로 변경
ViewController의 id를 관리하던 폴더를 삭제하고 StockNavigationController의 identifier를
통해서 접근하도록 변경

Commit:
refactor: StockViewController에 juiceMaker 캡슐화

StockViewController의 setJuiceMaker 함수를 구현해서 캡슐화 위배되는 문제를 해결

Commit:
docs: README 수정

Commit:
docs: README UML 추가

Commit:
style: 누락된 self 컨벤션 수정

jungseungyeo commented on JuiceMaker/JuiceMaker/Controller/StockViewController.swift:
```diff
@@ -13,7 +13,7 @@ final class StockViewController: UIViewController {
   
   var juiceMaker: JuiceMaker?
-  var delegate: StockDelegate?
+  weak var delegate: StockDelegate?
```
너무 좋습니다 :)

jungseungyeo commented on JuiceMaker/JuiceMaker/Controller/JuiceViewController.swift:
```diff
@@ -86,8 +86,8 @@ private extension JuiceViewController {
     }
-    stockViewController.juiceMaker = juiceMaker
     stockViewController.delegate = self
+    stockViewController.setJuiceMaker(juiceMaker)
```
캡슐화를 지키는 습관 좋습니다 :)

jungseungyeo reviewed (approved):
링고, 태엔젤 어느덧 마지막 한주가 지나갔네요

처음 pr을 보내던 것이 어그제 같은데 벌써 이렇게 쥬스메이커를 마무리가 되었네요. 그동안 너무 감사했습니다.
코드 수정을 보면서 더 이상 리뷰를 할 곳이 없네요 ㅎㅎ 

앞으로 더욱 성장하는 모습을 기대하고 있겠습니다. 고생하셨습니다.
